crypto/bio/bss_dgram_pair.c

```diff
@@ -922,22 +1036,25 @@ static int dgram_pair_recvmmsg(BIO *bio, BIO_MSG *msg,
     ossl_ssize_t l;
     BIO_MSG *m;
     size_t i;
-    struct bio_dgram_pair_st *b = bio->ptr, *peerb;
+    struct bio_dgram_pair_st *b = bio->ptr, *readb;
 
     if (num_msg == 0) {
         *num_processed = 0;
         return 1;
     }
 
-    if (b->peer == NULL) {
+    if (!bio->init) {
         ERR_raise(ERR_LIB_BIO, BIO_R_BROKEN_PIPE);
         *num_processed = 0;
         return 0;
     }
 
-    peerb = b->peer->ptr;
+    if (is_dgram_pair(b))
+        readb = b->peer->ptr;
+    else
+        readb = b;  
 
-    if (CRYPTO_THREAD_write_lock(peerb->lock) == 0) {
+    if (CRYPTO_THREAD_write_lock(readb->lock) == 0) {
```

Comment:
Is locking required for BIO_s_dgarm_mem?

Reply:
Yes - as above